test/testframe.c

```diff
@@ -11,368 +11,504 @@
-static int         num_errs                         = 0;         /* Total number of errors during testing */
-int                TestVerbosity                    = VERBO_DEF; /* Default Verbosity is Low */
-static int         Summary                          = 0;         /* Show test summary. Default is no. */
-static int         CleanUp                          = 1;         /* Do cleanup or not. Default is yes. */
-int                TestExpress                      = -1;   /* Do TestExpress or not. -1 means not set yet. */
-static TestStruct *Test                             = NULL; /* Array of tests */
-static unsigned    TestAlloc                        = 0;    /* Size of the Test array */
-static unsigned    Index                            = 0;
-static const void *Test_parameters                  = NULL;
-static const char *TestProgName                     = NULL;
-static void (*TestPrivateUsage)(void)               = NULL;
-static int (*TestPrivateParser)(int ac, char *av[]) = NULL;
-int mpi_rank_framework_g                            = 0;
+
+static TestStruct *TestArray = NULL; /* Array of tests */
+static unsigned    TestAlloc = 0;    /* Size of the Test array */
+static unsigned    TestCount = 0;    /* Number of tests currently added to test array */
+
+static const char *TestProgName                              = NULL;
+static void (*TestPrivateUsage_g)(FILE *stream)              = NULL;
+static herr_t (*TestPrivateParser_g)(int argc, char *argv[]) = NULL;
+
+static int  TestNumErrs_g        = 0;    /* Total number of errors that occurred for whole test program */
+static bool TestEnableErrorStack = true; /* Whether to show error stacks from the library */
+
+static int TestMaxNumThreads_g = -1; /* Max number of threads that can be spawned */
+
+static bool TestDoSummary_g = false; /* Show test summary. Default is no. */
+static bool TestDoCleanUp_g = true;  /* Do cleanup or not. Default is yes. */
+
+int TestFrameworkProcessID_g = 0;         /* MPI process rank value for parallel tests */
+int TestVerbosity_g          = VERBO_DEF; /* Default Verbosity is Low */
 
 /*
- * Setup a test function and add it to the list of tests.
- *      It must have no parameters and returns void.
- * TheName--short test name.
- *    If the name starts with '-', do not run it by default.
- * TheCall--the test routine.
- * Cleanup--the cleanup routine for the test.
- * TheDescr--Long description of the test.
- * Parameters--pointer to extra parameters. Use NULL if none used.
- *    Since only the pointer is copied, the contents should not change.
- * Return: Void
- *    exit EXIT_FAILURE if error is encountered.
+ * Add a new test to the list of tests to be executed
  */
-void
-AddTest(const char *TheName, void (*TheCall)(void), void (*Cleanup)(void), const char *TheDescr,
-        const void *Parameters)
+herr_t
+AddTest(const char *TestName, void (*TestFunc)(const void *), void (*TestSetupFunc)(void *),
```

Comment:
Added a test setup callback and a parameter for the size of test data so that the framework can copy the data and the same structure can be changed between calls to `AddTest()`